syntax = "proto3";

package TW.IoTeX.Proto;
option java_package = "wallet.core.jni.proto";

message Transfer {
    string amount  = 1;
    string recipient = 2;
    bytes payload  = 3;
}

message StakeCreate {
  string candidateName = 1;
  string stakedAmount = 2;
  uint32 stakedDuration = 3;
  bool autoStake = 4;
  bytes payload = 5;
}

message StakeReclaim {
  uint64 bucketIndex = 1;
  bytes payload = 2;
}

message StakeAddDeposit {
  uint64 bucketIndex = 1;
  string amount = 2;
  bytes payload = 3;
}

message StakeRestake {
  uint64 bucketIndex = 1;
  uint32 stakedDuration = 2;
  bool autoStake = 3;
  bytes payload = 4;
}

message StakeMove {
  uint64 bucketIndex = 1;
  string name = 2;
  bytes payload = 3;
}

message ContractCall {
    string amount  = 1;
    string contract = 2;
    bytes data = 3;
}

// transaction signing input
message SigningInput {
    uint32 version = 1;
    uint64 nonce = 2;
    uint64 gasLimit = 3;
    string gasPrice = 4;
    bytes privateKey = 5;
    oneof action {
        Transfer transfer = 10;
        ContractCall call = 12;
        StakeCreate stakeCreate = 40;
        StakeReclaim stakeUnstake = 41;
        StakeReclaim stakeWithdraw = 42;
        StakeAddDeposit stakeAddDeposit = 43;
        StakeRestake stakeRestake = 44;
        StakeChangeCandidate stakeChangeCandidate = 45;
        StakeTransferOwnership stakeTransferOwnership = 46;
        CandidateRegister candidateRegister = 47;
        CandidateBasicInfo candidateUpdate = 48;
  }
}

// transaction signing output
message SigningOutput {
    // Signed and encoded Action bytes
    bytes encoded = 1;

    // Signed Action hash
    bytes hash = 2;
}
